test/cppsuite/test_harness/workload/database_operation.h

```diff
@@ -49,164 +54,231 @@ class database_operation {
-                if (using_timestamps)
-                    tc->transaction.set_commit_timestamp(
-                      tc->session, timestamp_manager::decimal_to_hex(ts));
+            /* Start a transaction. */
+            if (!tc->transaction.active())
+                tc->transaction.begin(tc->session, "");
 
-                update(tc->tracking, cursors[i], collection_names[i], key.c_str(),
-                  generated_value.c_str(), ts);
+            /* Call next to pick a new random record. */
+            ret = cursor->next(cursor);
+            if (ret == WT_NOTFOUND)
+                continue;
+            else if (ret != 0)
+                testutil_die(ret, "unhandled error returned by cursor->next()");
 
-                /* Commit the current transaction. */
-                tc->transaction.op_count++;
-                if (tc->transaction.can_commit())
-                    tc->transaction.commit(tc->session, "");
-            }
-            tc->sleep();
+            /* Get the record's key. */
+            testutil_check(cursor->get_key(cursor, &key_tmp));
+
+            /*
+             * The retrieved key needs to be passed inside the update function. However, the update
+             * API doesn't guarantee our buffer will still be valid once it is called, as such we
+             * copy the buffer and then pass it into the API.
+             */
+            key = key_value_t(key_tmp);
+
+            /* Generate a new value for the record. */
+            generated_value =
+              random_generator::random_generator::instance().generate_string(tc->value_size);
+
+            /*
+             * Get a timestamp to apply to the update. We still do this even if timestamps aren't
+             * enabled as it will return WT_TS_NONE, which is then inserted into the tracking table.
+             */
+            ts = tc->timestamp_manager->get_next_ts();
+            if (using_timestamps)
+                tc->transaction.set_commit_timestamp(
+                  tc->session, timestamp_manager::decimal_to_hex(ts));
+
+            /*
+             * Update the record but take care to handle WT_ROLLBACK as we may conflict with another
+             * running transaction. Here we call the pre-defined wrappers as they also update the
+             * tracking table, which is later used for validation.
+             *
+             * Additionally first get the update_cursor.
+             */
+            cursor = collections[collection_id].update_cursor;
+            ret = update(tc->tracking, cursor, collections[collection_id].collection_name,
+              key.c_str(), generated_value.c_str(), ts);
+
+            /* Increment the current op count for the current transaction. */
+            tc->transaction.op_count++;
+
+            /* If the wiredtiger API has returned rollback, comply. */
+            if (ret == WT_ROLLBACK)
+                tc->transaction.rollback(tc->session, "");
```

Comment:
Should we add a note here saying that we should remove the update from the tracking table ?
I know it wont be done in this PR but we could leave a note.

Reply:
5abf1d1c6